src/borrow_check/region_inference.md

```diff
@@ -83,11 +87,86 @@ The kinds of region elements are as follows:
   for details on placeholders, see the section
   [placeholders and universes](#placeholder).
 
+## Constraints
+
+Before we can infer the value of regions, we need to collect constraints on the
+regions. There are two primary types of constraints.
+
+1. Outlives constraints. These are constraints that one region outlives another
+   (e.g. `'a: 'b`). Outlives constraints are generated by the [MIR type
+   checker].
+2. Liveness constraints. Each region needs to be live at points where it can be
+   used. These constraints are collected by [`generate_constraints`].
+
+[`generate_constraints`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/borrow_check/nll/constraint_generation/fn.generate_constraints.html
+
+## Inference Overview
+
+So how do we compute the contents of a region? This process is called _region
+inference_. The high-level idea is pretty simple, but there are some details we
+need to take care of.
+
+The [`RegionInferenceContext`] type contains all of the information needed to
+do inference, including the universal regions from `replace_regions_in_mir` and
+the constraints computed for each region. It is constructed just after we
+compute the liveness constraints.
+
+Here are some of the fields of the struct:
+
+- `constraints`: contains all the outlives constraints.
+- `liveness_constraints`: contains all the liveness constraints.
+- `universal_regions`: contains the `UniversalRegions` returned by
+  `replace_regions_in_mir`.
+- `universal_region_relations`: contains relations known to be true about
+  universal regions. For example, if we have a where clause that `'a: 'b`, that
+  relation is assumed to be true while borrow checking the implementation (it
+  is checked at the caller), so `universal_region_relations` would contain `'a:
+  'b`.
+- `type_tests`: contains some constraints on types that we must check after
+  inference (e.g. `T: 'a`).
+- `closure_bounds_mapping`: used for propagating region constraints from
+  closures back out to the creater of the closure.
+
+TODO: should we discuss any of the others fields? What about the SCCs?
```

Comment:
I feel like overall this discussion is a bit too focused on the "low-level details" and not much about the high-level process, but it seems like a good start.